synapse/storage/databases/main/event_push_actions.py

```diff
@@ -309,7 +309,45 @@ def add_thread_id_txn(
 
             # Update progress.
             processed_rows = txn.rowcount
-            progress[f"max_{table_name}_stream_ordering"] = max_stream_ordering
+            progress["max_event_push_actions_stream_ordering"] = max_stream_ordering
+            self.db_pool.updates._background_update_progress_txn(
+                txn, "event_push_backfill_thread_id", progress
+            )
+
+            return processed_rows
+
+        def add_thread_id_summary_txn(txn: LoggingTransaction) -> int:
+            min_user_id = progress.get("max_summary_user_id", "")
+            min_room_id = progress.get("max_summary_room_id", "")
+
+            sql = """
+            SELECT user_id, room_id FROM event_push_summary
+            WHERE (user_id, room_id) > (?, ?)
+                AND thread_id IS NULL
+            ORDER BY user_id, room_id
+            LIMIT 1
+            OFFSET ?
+            """
+
+            txn.execute(sql, (min_user_id, min_room_id, batch_size))
+            row = txn.fetchone()
+            if not row:
+                return 0
+
```

Comment:
Oh, actually, there's an issue with this because of the OFFSET. If you don't have a number of rows divisible by the batch size, we'll get some rows 'left over'. You'd have to do a final pass with `> (min_user_id, min_room_id)`? :/

Alternatively: do an `UPDATE .. RETURNING` and use `rows[-1]` as the next min

Reply:
I thought about different ways of doing this, in the end i just made it not include an upper bound if we don't have an upper bound.